cp23/hpi-fun2/comppong/executables/generator.cpp

```diff
@@ -0,0 +1,137 @@
+#include <bits/stdc++.h>
+#include <testlib.h>
+
+using namespace std;
+
+const string_view SAMPLE2 = R"(5 4
+2  3  1  4 -2
+-2  2 -4 -1  2
+-3 -5 -1  6 -5
+-1 -2  2  6  9
+)";
+
+const string_view SAMPLE1 = R"(3 3
+-2 -1 -1
+1 -1 1
+1 2 -2
+)";
+
+const string_view SAMPLE3 = R"(1 8
+-3
+4
+-1
+-5
+2
+5
+-1
+2
+)";
+
+const string_view MULTIPLESOLUTIONS = R"(3 1
+1 -1 1
+)";
+
+const string_view MULTIPLESOLUTIONS2 = R"(3 1
+1 -3 1
+)";
+
+const string_view WHOLEFIELD = R"(4 4
+1 1 1 1 
+1 -1 -1 1
+1 -1 -1 1
+1 1 1 1
+)";
+
+const string_view SINGLECUP = R"(2 2
+1 -1
+-1 -1
+)";
+
+template <class F>
+void testcase(string name, string desc, F f)
+{
+  ofstream desc_file(name + ".desc");
+  desc_file << desc;
+  string in_file = name + ".in";
+  freopen(in_file.c_str(), "w", stdout);
+  f();
+}
+
+void predefined(string name, string desc, string_view content)
+{
+  testcase(name, desc, [&]()
+           { cout << content; });
+}
+
+void sample(int num, string_view content)
+{
+  auto num_str = toString(num);
+  predefined("sample" + num_str, "Sample #" + num_str, content);
+}
+
+int main(int argc, char *argv[])
+{
+  registerGen(argc, argv, 1);
+  rnd.setSeed(5120471233170915017ll);
+
+  sample(1, SAMPLE1);
+  sample(2, SAMPLE2);
+  sample(3, SAMPLE3);
+  predefined("multiple_solutions", "multiple_solutions", MULTIPLESOLUTIONS);
+  predefined("multiple_solutions2", "multiple_solutions2", MULTIPLESOLUTIONS2);
+  predefined("whole_field", "whole_field", WHOLEFIELD);
+  predefined("single_cup", "single_cup", SINGLECUP);
+
+  testcase("random_nonquadratic", "random_nonquadratic", []()
+           {
+    int w = 20;
+    int h = 25;
+    cout << w << " " << h << endl;
+    for (int i = 0; i < h; i++) {
+      for (int j = 0; j < w; j++) {
+        if (i == 0 && j == 0)
+          cout << 1 << ' ';
+        else if (i == h - 1 && j == w - 1)
+          cout << -1 << ' ';
+        else
+          cout << rnd.next(-100, 100) << ' ';
+      }
+      cout << endl;
+    } });
+
+  testcase("random_quadratic", "random_quadratic", []()
+           {
+    int w = 15;
+    int h = 15;
+    cout << w << " " << h << endl;
+    for (int i = 0; i < h; i++) {
+      for (int j = 0; j < w; j++) {
+        if (i == 0 && j == 0)
+          cout << 1 << ' ';
+        else if (i == h - 1 && j == w - 1)
+          cout << -1 << ' ';
+        else
+          cout << rnd.next(-100, 100) << ' ';
+      }
+      cout << endl;
+    } });
+
+  testcase("random_row", "random_row", []()
+           {
+    int w = 100;
+    int h = 1;
+    cout << w << " " << h << endl;
+    for (int i = 0; i < h; i++) {
+      for (int j = 0; j < w; j++) {
+        if (i == 0 && j == 0)
+          cout << 1 << ' ';
+        else if (i == h - 1 && j == w - 1)
+          cout << -1 << ' ';
+        else
+          cout << rnd.next(-100, 100) << ' ';
+      }
+      cout << endl;
+    } });
+
+  return 0;
+}
```

Comment:
This violates your current bounds on the input. Though I do agree that it would be nice to bound only the sum of $w, h$ and require the implementation to take advantage and be in $\mathcal O(\min(w,h)^2 \cdot \max(w,h))$ rather than the opposite.

Reply:
I would also expect there to be a column testcase as well.

Reply:
There is no maximum testcase as of right now.

Reply:
You should look into refactoring this code as all of your random testcases could be expressed as a call to a function with parameters `w` and `h`